Kubernetes ContainerFactoryProvider implementation

jcrossley3 opened the pull request:
We've had this working in a fork for a few months now, but have held off submitting due to performance limits imposed by our inability to `suspend/resume` containers like the Docker provider. However, @dgrove-oss has some ideas around this and asked us to go ahead and open the PR.

One other limitation is that the k8s API doesn't distinguish between `stdout` and `stderr` for logged messages. So in our impl, all messages (even those sent to `stderr`) appear to go to `stdout`. See https://github.com/kubernetes/kubernetes/issues/28167

I squashed a bunch of commits for this but wanted to maintain a bit of history for posterity. If you'd prefer a single commit, that's cool, too.

dgrove-oss commented on core/invoker/Dockerfile:
```diff
@@ -11,6 +12,11 @@ rm -f docker-${DOCKER_VERSION}.tgz && \
+RUN wget --no-verbose https://storage.googleapis.com/kubernetes-release/release/v${KUBERNETES_VERSION}/bin/linux/amd64/kubectl && \
+chmod +x kubectl && \
+mv kubectl /usr/bin/kubectl
+
```
I was expecting the client would talk directly to Kubernetes's API service using its REST API, not use kubectl.   I see that the KubernetesContainer has a mix of REST calls using the fabric.io client and kubectl commands.  Is due to ease of development, or is there a performance/functional reason to use kubectl to do the run command?

bbrowning replied:
This is just an artifact of copying the original Docker implementation that shelled out to `docker`. We had to replace usage of `kubectl` with the API for performance reasons in some places but it looks like we never got around to replacing it everywhere.

dgrove-oss replied:
makes sense; thanks.

dgrove-oss commented:
Happy to report this worked for me (once I figured out the one magic config variable to pass on the command line for my kube environment...).  I have a WIP PR in the deploy-kube that adds the kubernetes yml files to deploy an invoker that uses this new pool.

rabbah commented on core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala:
```diff
@@ -37,7 +37,7 @@ import akka.stream.scaladsl.{Framing, Source}
 import spray.json._
-import whisk.core.containerpool.logging.LogLine
+import whisk.core.containerpool.logging.{LogLine => CoreLogLine}
 import whisk.http.Messages
```
Why this import remapping?

rabbah commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala:
```diff
@@ -0,0 +1,222 @@
+        name
+      }
+      .map(ContainerId.apply)
+  }
```
Why not flatten the two maps to:

.map(_ => ContainerId(name))

rabbah commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala:
```diff
@@ -0,0 +1,222 @@
+
+  private def runCmd(args: Seq[String], timeout: Duration)(implicit transid: TransactionId): Future[String] = {
+    val cmd = kubectlCmd ++ args
+    val start = transid.started(
```
Recent changes to this logger method means this is only emitted at debug level. FYI.

jcrossley3 replied:
Looks like `DockerContainer` explicitly sets `InfoLevel` now. I'll update to match.

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesContainer.scala:
```diff
@@ -0,0 +1,159 @@
+  /** The last read timestamp in the log file */
+  private val lastTimestamp = new AtomicReference("")
+
+  protected val logsRetryCount = 15
```
No longer used.

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesContainer.scala:
```diff
@@ -0,0 +1,159 @@
+  private val lastTimestamp = new AtomicReference("")
+
+  protected val logsRetryCount = 15
+  protected val logsRetryWait = 100.millis
```
No longer used either.

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala:
```diff
@@ -0,0 +1,222 @@
+
+  def inspectIPAddress(id: ContainerId)(implicit transid: TransactionId): Future[ContainerAddress] = {
+    Future {
+      blocking {
```
Is there a non-blocking client prodiving the features we need? (https://github.com/doriordan/skuber for example?)

dgrove-oss replied:
All the clients are going to bottom out with making http put/get against the same k8s REST API.  So, I'm not sure if it really matters in the end if we have the blocking explicit in our code or hide it in the bowels of 3rd party client instead.  

I'd bias towards picking a client based on expected long-term maintenance/support, expectation of it tracking k8s API changes, and license compatibility.

markusthoemmes replied:
Sure, that's why I added "providing the features we need". I'm all in for not just picking it for the sake of it. Still a non-blocking client is preferable over a blocking one (if it exists). This might matter if you want to scale the invoker to more than the 16 containers that we scale it today.

bbrowning replied:
We're planning to scale each Invoker to way more than 16 containers, but we haven't done large-scale load testing of this yet. I'd say we're targeting at least 100 or more containers per Invoker.

dgrove-oss replied:
Circling back on the client question.  The fabric8io client appears to lack support for specifying pod affinity.  Skimming through the source code, I see no support for affinity (and several open issues on GitHub asking for affinity support). The skuber client looks like it does have affinity support.  I think affinity is a required feature (lets us tell kube to restrict creation of user action containers to designated invoker nodes).

bwmcadams replied:
With regards to the non-blocking, I've solved this down stream and am working on finishing testing. The fabric8 uses okhttp, which has an async mode. I've migrated to a custom `Source`, using the async callbacks, to eliminate most of the cruft and any blocking code. (I'm not sure about the pod affinity however)

dgrove-oss replied:
cool!  Maybe we can help out by adding the pod affinity support downstream as well.  I don't think it should be too hard.

bwmcadams replied:
After reviewing, with regards to this section of code, I can't see a way with the fabric8 java client to *not* block here. Even if you avoid the utility method for waiting that we use in this code, every layer all the way down the client blocks at some point for it. and since this isn't able to just be a simple REST call like the log fetch, I can't refactor it to be async. The blocking is wrapped and marked inside the future, however, and will be handled safely.

dgrove-oss replied:
Makes sense to me.  Fundamentally we need to wait somehow until the pod is ready before we can inspect its Status object to get the PodIP and similar info.

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala:
```diff
@@ -0,0 +1,222 @@
+            "pods",
+            id.asString,
+            "log?timestamps=true" ++ sinceTimePart))
+        val request = new Request.Builder().get().url(url).build
```
Would it make sense to use Akka's URI building capabilities here? Could get much more readable.

```scala
val sinceTime = Some("abc")

val path = Path / "api" / "v1" / "namespaces" / "namespace" / "pods" / "pod"
val query = Seq("timestamp" -> true.toString) ++ sinceTime.map(time => "sinceTime" -> time)
val url = Uri("https://google.de")
  .withPath(path)
  .withQuery(Query(query: _*))
```

That assumes that `sinceTime` is an `Option[String]` to denote the possible absence of it in the type. In general, it'll make your code much easier to grasp if you make use of `Option` whenever it is possible that something is missing rather than passing an empty String and checking for it to be empty.

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala:
```diff
@@ -0,0 +1,222 @@
+        val response = kubeRestClient.getHttpClient.newCall(request).execute
+        if (!response.isSuccessful) {
+          Future.failed(
+            new Exception(s"Kubernetes API returned HTTP status ${response.code} when trying to retrieve pod logs"))
```
This failed `Future` will be thrown away and not used anywhere. I think what you're looking for is:

```scala
Future {
  blocking {
    // your request code
    val response = kubeRestClient.getHttpClient.newCall(request).execute
    if(response.isSuccessful) {
      Future.successful(response.body.string)
    } else {
      Future.failed(
            new Exception(s"Kubernetes API returned HTTP status ${response.code} when trying to retrieve pod logs"))
    }
  }
}.flatMap(identity).map { output => ... }
```

An `if` is an expression in Scala, meaning it'll return something (namely: the value of the last line). The `Future { }` block in your case always returns `response.body.string`. If you want to fold the error case into the future chain, you'll explicitly need to return a `Success` for the good case and a `Failure` for the bad case. Since it makes chaining easier, I used the `Future.successful` and `Future.failed` variants, which are immediatly resolved Futures of the desired state. The return type would now be `Future[Future[String]]` though, since I returned a Future inside another Future. To flatten those (flatten is only available in Scala 2.12) you can use `flatMap(identity)`. `identity` is a predefined function. You can as well write `flatMap(x => x)`.

Hope that makes sense.

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala:
```diff
@@ -0,0 +1,222 @@
+    } match {
+      case Seq() => Seq()
+      case pairs => Seq("-l") ++ pairs
+    }
```
Do you need to add `"-l"` in front of every pair or just in front of all pairs? Currently it's the latter.

jcrossley3 replied:
When selecting by multiple labels, you can either use multiple `"-l"`flags or a single `"-l"` with a comma-delimited list. The logic for the former is "any can match" and the logic for the latter is "all must match". At least I think that's how it works. ;)

Right now, I believe we only ever pass one label anyway, so I'm inclined to leave it as is for now.

dgrove-oss replied:
fwiw, I'm prototyping a version where we create the pod for the user action container using the client (not kubectl).  That eliminates this code entirely.

markusthoemmes replied:
If we only want this to work for one label, we should change `labels: Map[String, String]` to `label: (String, String)` or make the code work for multiple labels. Leaving a known bug seems odd to me and might become frustrating for others who want to enhance/refactor that code.

jcrossley3 replied:
I wouldn't call it a bug. The code does work for multiple labels, in the way I would expect if I did pass multiple labels. I get what you're saying about changing the code to work for one label since that's how we're currently using it, but to me that conflicts with how one might generically use `kubectl` itself. I don't feel that strongly about it, though, especially in light of @dgrove-oss attempt to obviate it. :)

dgrove-oss replied:
This code would become something like: https://github.com/dgrove-oss/openwhisk/commit/7a72e1913b7e8a3e49f6ccfd5c69441fc90801db

I'm debating between constructing the pod like this (via code) vs. loading a template .yml file and then filling in a few holes (with limit, env, name).  I have a guess that this is a place where providers may want to customize, but maybe we should defer that decision until later.

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala:
```diff
@@ -0,0 +1,222 @@
+ */
+class KubernetesClient(
+  timeouts: KubernetesClientTimeoutConfig = loadConfigOrThrow[KubernetesClientTimeoutConfig](
+    "whisk.kubernetes.timeouts"))(executionContext: ExecutionContext)(implicit log: Logging, as: ActorSystem)
```
Please externalize the config key to `ConfigKeys`

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesContainer.scala:
```diff
@@ -0,0 +1,159 @@
+          .parseJson
+          .asJsObject
+          .fields("time")
+          .convertTo[String]
```
It's not necessary to match on lines and non-emptyness here, since we have successfully passed the Framing-delimiter we know we only have 1 line at hand.

What about:

```scala
lastTimestamp.set(bs.utf8String.convertTo[LogLine].time)
```

Parsing to `LogLine` should be safe since you created the objects accordingly above.

jcrossley3 replied:
Very cool!

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala:
```diff
@@ -0,0 +1,222 @@
+            val msg = line.substring(pos + 1)
+            // TODO - can we get the proper stream name from kubernetes? Some stuff is stderr
+            val stream = "stdout"
+            JsObject("log" -> msg.toJson, "stream" -> stream.toJson, "time" -> ts.toJson)
```
What about using `LogLine(msg, stream, ts).toJson.compactPrint` instead? That would ensure the proper format and fieldnames.

jcrossley3 replied:
I don't seem to have those methods on `LogLine`. Is it enough to just `LogLine(msg, stream, ts)`?

markusthoemmes replied:
`LogLine` is a case-class, you'll likely need to import it. But then, yes, it should just be that.

jcrossley3 replied:
One of us may be confused. Just to make sure you know what this code is doing: we are adapting the k8s log output, making it look like the docker log output, so that we can re-use the docker classes in `KubernetesContainer.logs`. I concede it's a hack, but I think it means we can't use `LogLine` at this point.

markusthoemmes replied:
Right I'm aware. Using `LogLine` just **ensures** you're adapting correctly (because it's used for deserialization again). Both you're any my line should yield the same result, but using `LogLine` ensures you're using the right field names etc.

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala:
```diff
@@ -0,0 +1,222 @@
+      }
+    }.map { output =>
+      val original = output.lines.toSeq
+      val relevant = original.dropWhile(s => !s.startsWith(sinceTime))
```
Can you elaborate a bit on this piece? I don't quite grasp why you're dropping lines having the exact same timestamp as the sinceTime and then decide based on that what to keep. That is to say: A comment describing it would be good.

dgrove-oss commented:
PG1 / 2574 🏃

dgrove-oss commented:
@jcrossley3 - I dug into the PG test failures this morning.  The problem is that the PR needs to be rebased to pick up the change from #3063  that moves common/scala/Dockerfile to use alpine linux as a base image.  Without that change in base image, it won't go through the PG testing pipeline.

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala:
```diff
@@ -0,0 +1,218 @@
+            val stream = "stdout"
+            LogLine(ts, stream, msg).toJson.compactPrint
+          }
+          .mkString("\n") + "\n" // trailing newline is necessary or the frame won't be decoded and break the akka stream
```
Why not just add the `"\n"` on the line above:

```scala
LogLine(ts, stream, msg).toJson.compactPrint + "\n"
```

jcrossley3 replied:
Makes sense to me, I think. :)

bwmcadams replied:
the "why" of the trailing newline is my fault; no logic to it and the suggested adjustment makes sense. 

If I'm understanding you correctly, @markusthoemmes… right now we're splitting the entire logging result based on the HTTP response, and pushing it out as a single item. Your suggestion is to break it out so the stream takes individual events for each line, yes?

markusthoemmes replied:
@bwmcadams Yep, it goes through a framing stage immediatley anyway so I don't expect much difference in either way. It just feels clearer to have it immediately exploded into the specific events and squash them into the stream instead of creating a string just to split it immediatly.

markusthoemmes replied:
Actually, thinking about it: If you choose to implement my proposal you can just drop the FramingStage below.

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala:
```diff
@@ -0,0 +1,218 @@
+          relevant.drop(1)
+        } else {
+          original
+        }
```
Hmm, I'm still not able to grasp this, let me try:

```scala
val original = output.lines
val relevant = sinceTime.map(time => original.dropWhile(!_.startsWith(time)).drop(1)).getOrElse(original)
```

`relevant` should no in any case contain what you need to continue processing, right? The following `if` is redundant, because if relevant is empty, the result should be empty as well?

jcrossley3 replied:
If `relevant` is empty, it means the previous timestamp isn't in `original`. This will be the case for both cold pods and warm pods that haven't been used within the granularity of the timestamp.

jcrossley3 replied:
I should clarify about the warm pods. It's not necessarily that they haven't been used. It's that the log messages from the previous activation are within the server's granularity relative to the activation's last message. This is very rarely more than just the two stream sentinel markers.

jcrossley3 replied:
When I read my comments, I don't feel like they're helping you. :)

bwmcadams replied:
@markusthoemmes consider most of this thread obsolete; I've done a rewrite with all of this handled as a custom `GraphStage` and it's much saner. It cuts out most of the hackiness of the existing code.

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala:
```diff
@@ -0,0 +1,218 @@
+      "--restart",
+      "Always",
+      "--limits",
+      "memory=256Mi") ++ environmentArgs ++ labelArgs
```
1. Looks to me like `memory=256Mi` should be passed in from the outside.
2. Why `restart Always`? For user-containers?
3. Do we need to make the generator configurable?

jcrossley3 replied:
#1 makes sense. I'm gonna refer to @bbrowning for #2 and #3

bbrowning replied:
This may be specific to OpenShift, but the reason for this is so that these pods don't get treated as temporary pods with different lifecycle semantics, sometimes resulting in them being prematurely killed. By setting `restart` to `Always` we prevent them from being treated as temporary pods. However, if you only set that, then instead of creating a Pod object it ends up creating a replication controller. So, we also pass the `generator` to ensure it just creates a Pod.

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesContainerFactory.scala:
```diff
@@ -0,0 +1,81 @@
+      userProvidedImage = userProvidedImage,
+      environment = Map("__OW_API_HOST" -> config.wskApiHost),
+      labels = Map("invoker" -> label),
+      name = Some(name))
```
Memory usage should be passed in here?

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesContainer.scala:
```diff
@@ -0,0 +1,145 @@
+      .via(Framing.delimiter(delimiter, limit.toBytes.toInt))
+      .via(new CompleteAfterOccurrences(activationMarkerCheck, 2, waitForSentinel))
+      .recover {
+        case _: StreamLimitReachedException =>
```
Is this Exception relevant? You dropped the limitWeighted stage. Was that deliberate? You'll read/store unbounded data.

jcrossley3 commented:
Ok, PR has been rebased against current master and unit tests have been added. There are 2 pending ones in there that revealed some issues @bwmcadams is in the process of addressing. Yay, unit tests! :)

rabbah commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesContainer.scala:
```diff
@@ -0,0 +1,165 @@
+
+    for {
+      id <- kubernetes.run(podName, image, args).recoverWith {
+        case _ => Future.failed(WhiskContainerStartupError(s"Failed to run container with image '${image}'."))
```
We recently (@markusthoemmes specifically) changes the content of this message to use
```
  val resourceProvisionError = "Failed to provision resources to run the action."
```
since the image name may be mangled internally for some images.

Perhaps we should use that here.
@markusthoemmes we should may just make that the default so we prevent this in the future from possibly leaking out to activation responses.

rabbah replied:
Same for line 99 below.

jcrossley3 replied:
Done

rabbah commented:
@jcrossley3 I know you want to preserve the history in this PR so when @markusthoemmes approves/dismisses his review, can you adjust the last few commit and stash them appropriately? One of us can merge without squashing the full history.

markusthoemmes commented:
Regarding credits for all committers: Github introduced https://github.com/blog/2496-commit-together-with-co-authors. Would you be okay using that and still squash down to one commit?

rabbah commented:
nice! thanks for sharing it, I hadn't see that yet.

jcrossley3 commented:
@rabbah do you want all of this PR's commits squashed into a single one? Or just those pushed in response to feedback?

rabbah commented:
With the git enhancement linked by Markus you don’t need to squash at all. We can just amend the commit message when merging.

markusthoemmes commented:
@jcrossley3 since all feedback seems to be addressed, it'd still be good if you can squash down or at least ammend the first commit in a way to conform with the feature I linked. It seems to be a bit cumbersome to add the proper email adresses and everything. You'll have an easier time doing that than the reviewers/mergers.

jcrossley3 commented:
@markusthoemmes I added more tests around the log handling and implemented your code suggestions in `KubernetesClient` which @bwmcadams is in the midst of changing as well. When he pushes his fixes, I'll squash the most recent commits with the email addresses.

markusthoemmes commented:
@jcrossley3 You'll need to rebase. Are all commit from @bwmcadams in?

jcrossley3 commented:
Not quite yet @markusthoemmes. In fixing the tests, @bwmcadams took it upon himself to refactor things to remove the blocking calls. Once he pushes that change and you guys approve it, I'll rebase.

jcrossley3 commented:
Ok, we've rebased and changed the logging stuff quite a bit, but I think we have good unit test coverage now, at least parity with our Docker counterparts. Once you guys approve, let me know how much you'd like to squash. I say probably at least back to 3ef07bf and maybe even farther.

dgrove-oss commented:
Started PG1/2632 to see how things look.

dgrove-oss commented:
PG1/2632 looked clean (2 test failed; not specific to this PR (also failing on other PG runs).

dgrove-oss commented on tests/src/test/scala/whisk/core/containerpool/kubernetes/test/KubernetesClientTests.scala:
```diff
@@ -0,0 +1,272 @@
+  def kubernetesClient(fixture: => Future[String]) = new KubernetesClient()(global) {
+    override def findKubectlCmd() = kubectlCommand
+    override def executeProcess(args: Seq[String], timeout: Duration)(implicit ec: ExecutionContext, as: ActorSystem) =
+      fixture
```
fwiw, I don't think mocking out executeProcess is going to be a useful long-term tool for unit tests.  As we replace kubectl commands with calls to the kubernertes REST API, we lose executeProcess as a hook point.  For example, in a branch based off this PR, I've switched to doing pod creation via the REST API and as a result lost this hook point for all the "run" unit tests (so I disabled them).

jcrossley3 replied:
DISABLED THE TESTS WHAT!?!?! ;)
Yeah, @bwmcadams  had to set up dummy `Source` objects to validate the log parsing logic. I guess we'll need to figure out something similar as we convert the other calls to the API.

dgrove-oss replied:
open for suggestions on how to unit test run when it becomes just a thin veneer over the kubernetesClient API.  Code is sitting here if you want a preview: https://github.com/dgrove-oss/openwhisk/tree/k8s-invokeragent-20180219 (I've cleaned up & consolidated my older branches into 3 logical commits when I rebased them onto the latest code from this PR).

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala:
```diff
@@ -0,0 +1,324 @@
+    qB += "timestamps" → "true"
+    qB ++= sinceTime.map(time => "sinceTime" -> time.format(KubernetesClient.K8SDateTimeFormat))
+
+    Query(qB.result())
```
Why use a manual `Builder` here?

Why not:

```scala
Map("timestamps" -> "true") ++ sinceTime.map(time => "sinceTime" -> time.format(KubernetesClient.K8SDateTimeFormat))
```

bwmcadams replied:
Looks like that's an artifact that survived a refactor and didn't get cleaned up; will fix.

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala:
```diff
@@ -0,0 +1,324 @@
+    .appendLiteral('T')
+    .appendPattern("HH:mm:ss[.n]")
+    .appendLiteral('Z')
+    .toFormatter()
```
This looks like the default `Instant.toString` format. There's also `Instant.parse` for the inverse. Is a manual pattern needed here?

Can we change the whole code to use `Instant` instead of `LocalDateTime`?

bwmcadams replied:
It appears you are correct in the formatting; originally we had two split formatters for posting and reading because the REST endpoint doesn't accept nanos (at the time, I thought it just ignored the whole date, but turns out it just skips the nanos). Switching this to instant makes perfect sense.

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala:
```diff
@@ -0,0 +1,324 @@
+        stream = "stdout" // TODO - when we can distinguish stderr: https://github.com/kubernetes/kubernetes/issues/28167
+      } yield {
+        tsDate -> JsObject("log" -> msg.toJson, "stream" -> stream.toJson, "time" -> ts.toJson)
+      }) match {
```
Some more descriptive variable names might help here:

```
line vs. l
timestampDelimiter vs. p
timestamp vs. ts
parsedTimestamp vs. tsDate
```

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala:
```diff
@@ -0,0 +1,324 @@
+          readLines(src, lastTimestamp, lines)
+      }
+    } else {
+      src.close()
```
Should the caller be responsible of closing the source? Since it isn't opened by `readLines`

bwmcadams replied:
It struck me at the time as the sanest approach, since the HTTP library will drop lots of exceptions and warnings in  the logs if you don't close the source. But it isn't clean; i'll move it back out to the caller of `readLines`

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesContainer.scala:
```diff
@@ -0,0 +1,171 @@
+        Try(LocalDateTime.parse(parsed.time, K8SDateTimeFormat)).foreach(time => lastTimestamp.set(Option(time)))
+        true
+      } else false
+    }
```
Is it not feasible to alter the entire stream for the kube-case to avoid to reparse all the loglines just to set the latest timestamp?

You could, for instance, return a fully parsed object (`LogLine` or something like a new `TypedLogLine(Instant, Stream, String)`?) from `kubernetes.logs`, drop the Framing, adjust limitWeighted and then have a `.map(_.toJson.compactPrint)` as the very last stage to be pluggable into the logstores.

You're parsing everything anyway, why not keep it that way? Is it just test-parity?

markusthoemmes commented on tests/src/test/scala/whisk/core/containerpool/kubernetes/test/KubernetesClientTests.scala:
```diff
@@ -0,0 +1,272 @@
+        source
+          .via(Framing.delimiter(ByteString("\n"), 1024))
+          .via(Flow[ByteString].map(l => Try(l.utf8String.parseJson.convertTo[LogLine]).getOrElse(LogLine("", "", ""))))
+          .filter(_ != LogLine("", "", ""))
```
Why not `mapConcat(l => Try(l.utf8String.parseJson.convertTo[LogLine])` and drop the following filter?

markusthoemmes commented on tests/src/test/scala/whisk/core/containerpool/kubernetes/test/KubernetesContainerTests.scala:
```diff
@@ -0,0 +1,550 @@
+
+  implicit val materializer: ActorMaterializer = ActorMaterializer()
+
+  def instantDT(instant: Instant): LocalDateTime = LocalDateTime.from(instant.atZone(ZoneId.of("GMT+0")))
```
"UTC"?

bwmcadams replied:
UTC? This is one of those European things, like Celsius, and Kilometers, right?

glikson commented:
should have "kubernetes" label?
